### PR TITLE
Port storage/ to Python 3

### PR DESCRIPTION
**Pull request by hawkowl**

### CODE REVIEW

File: synapse/storage/_base.py [hunk 1]
```diff
@@ -1216,3 +1217,46 @@ class _RollbackButIsFineException(Exception):
...
+
+    # Decode it to a Unicode string before feeding it to json.loads, so we
+    # consistenty get a Unicode-containing object out.
+    if isinstance(db_content, (bytes, bytearray)):
+        db_content = db_content.decode('utf8')
+
+    try:
+        return json.loads(db_content)
+    except Exception:
+        # I don't know how, but somehow some binary objects get mangled upon
```
Comment (ara4n) [inline synapse/storage/_base.py]: I'm assuming this is happening in the weird tables where we inexplicably store JSON as binary, like `users_filter`.  Looking at this with psql, the rows look like:

```
 matthew |       130 | \x7b22726f6f6d223a7b2274696d656c696e65223a7b226c696d6974223a32307d7d7d
```

which sure enough looks like \x7b ({) followed by a bunch of ASCII hex digits.

So, either it's being inserted in the first place in the mangled format, or the postgres client library (or server?) is incorrectly displaying it with only a single preceding \x rather than correctly putting a \x before each pair of hex digits.

The type of this column is:

`filter_json | bytea`

Looks like postgres is behaving as intended when displaying a bytea as `\x.............`; this is what you get if the bytea_output global is set to `hex`.  If you set it to `escape` (the other valid value), you get this instead:

```
matrix=# set bytea_output='escape';
SET
matrix=# SHOW bytea_output;
 bytea_output 
--------------
 escape
(1 row)

matrix=# select * from matrix.user_filters where user_id='matthew' limit 1;
 user_id | filter_id |             filter_json              
---------+-----------+--------------------------------------
 matthew |         0 | {"room": {"timeline": {"limit": 8}}}
(1 row)
```

So it looks like the data is okay in the DB, it's just that the python postgres library is interpreting `\x1234` as a python-style hex escape sequence rather than this horrible mutant postgres hex sequence.

So, looking at psycopg's FAQ (http://initd.org/psycopg/docs/faq.html#faq-bytea-9-0), it turns out we're not the only people with these woes, and the solution is simply to do the same trick as above and explicitly `set bytea_output='escape';` and then everything will be green & submarine.

tl;dr: let's kill this with fire and `set bytea_output='escape';` instead :)

Comment (ara4n) [reply synapse/storage/_base.py]: hm, actually, this might be insufficient - looking at https://github.com/cockroachdb/cockroach/pull/17693/files where they had a similar problem, they ended up manually unescaping it much as you have in the app :/

Comment (ara4n) [reply synapse/storage/_base.py]: actually, reading the FAQ some more, it sounds like this shouldn't even be a bug if we are running psycopg 2.4.1 or later with libpq from 9.0 or later.  On matrix.org right now we look to be running psycopg 2.7.5 with libpq 5.10 (i.e. for postgres 10) - so i think this should be absolutely fine.  Was this just that you were testing on a box using an ancient python postgres client?

I guess we should probably `set bytea_output='escape'` to be safe.

Comment (ara4n) [reply synapse/storage/_base.py]: (even nicer would be to kill these bytea tables with fire - i have **no idea** why they are byteas; it looks like they were originally created as LONGBLOBs in a fit of overenthusiasm/confusion, despite them just being utf8 JSON)

Comment (hawkowl) [reply synapse/storage/_base.py]: @ara4n I noticed this coming out of *your* homeserver :)

Comment (erikjohnston) [reply synapse/storage/_base.py]: We should probably pin psycopg2 (and libpq, whose version psycopg2 helpfully exports). Debian jessie have suitably up to date packages, so it should be safe.

File: synapse/storage/events.py [hunk 2]
```diff
@@ -1908,9 +1908,9 @@ def _purge_history_txn(
             (room_id,)
         )
         rows = txn.fetchall()
-        max_depth = max(row[0] for row in rows)
+        max_depth = max(row[1] for row in rows)
 
-        if max_depth <= token.topological:
+        if max_depth < token.topological:
```
Comment (erikjohnston) [inline synapse/storage/events.py]: I don't think we want to remove the equality check here.

File: synapse/storage/events_worker.py [hunk 3]
```diff
@@ -12,6 +12,9 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
+from __future__ import division
```
Comment (erikjohnston) [inline synapse/storage/events_worker.py]: Is this necessary given we seem to only be doing `//` in here?

File: synapse/storage/events.py (hunk 2 above)
Comment (hawkowl) [reply synapse/storage/events.py]: it does, because otherwise the tests fail :(

File: synapse/storage/events_worker.py (hunk 3 above)
Comment (hawkowl) [reply synapse/storage/events_worker.py]: maybe not, I don't think

File: synapse/storage/engines/postgres.py [hunk 4]
```diff
@@ -47,6 +60,7 @@ def on_new_connection(self, db_conn):
         if not self.synchronous_commit:
             cursor = db_conn.cursor()
             cursor.execute("SET synchronous_commit TO OFF")
+            cursor.execute("SET bytea_output TO escape")
```
Comment (krombel) [inline synapse/storage/engines/postgres.py]: This does not look like the right place for it.
`synchronous_commit` defaults to `true` so this will in default not do that what you expect it to do.

Comment (hawkowl) [reply synapse/storage/engines/postgres.py]: wait, shit. that's right. thanks for catching it, let me fix that up.
